src/dune_pkg/flock.ml

```diff
@@ -0,0 +1,75 @@
+open! Stdune
+module Flock = Dune_util.Flock
+module Scheduler = Dune_engine.Scheduler
+
+let rec attempt_to_lock flock lock ~timeout_s =
+  let open Fiber.O in
+  match Flock.lock_non_block flock lock with
+  | Error e -> Fiber.return @@ Error e
+  | Ok `Success -> Fiber.return (Ok `Success)
+  | Ok `Failure ->
+    let sleep_duration_s = 0.1 in
```

Comment:
This could possibly be pulled out of the function. I don't know, even in the original code it is an admittedly ugly bit.

Reply:
I've split this into two mutually recursive closures to avoid needing to pass as much info from one recursive call to the next.